File: apps/api/app/schemas/csv_import.py
"""Request/response shapes for CSV import (plan section 13.1) — one generic
pipeline for any page: upload -> preview (encoding/delimiter/header +
suggested mapping) -> validate (every row, against the live schema, with
natural-key dedup) -> commit (chunked insert + `import_batches` row) ->
optional rollback within 24 hours.

Owner only, same as export. Every step re-sends the file: there is no
server-side session/cache for a partially-completed import (no such
infrastructure exists yet, and re-parsing a <=10MB CSV is cheap), so the
client holds the picked file in memory across the wizard's steps.
"""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

#: How to read an ambiguous (slash-separated) date/datetime column — the
#: Owner sets this once per column, not per cell (plan section 13.1:
#: "ambiguous dates prompt the Owner to confirm the format. Never guess.").
#: Unambiguous ISO text (`2026-09-01`) is always accepted regardless of this.
DateFormatHint = Literal["YMD", "DMY", "MDY"]

DuplicateStrategy = Literal["skip", "update", "create_anyway"]


class ImportRowError(BaseModel):
    #: 1-indexed data row (the header doesn't count as row 1).
    row: int
    column: str | None = None
    message: str


class ImportPreviewResponse(BaseModel):
    encoding: str
    delimiter: str
    headers: list[str]
    #: CSV header -> this page's column `key`, fuzzy-matched by name; `null`
    #: for a header with no confident match. FORMULA/ATTACHMENT columns and
    #: this page's generated columns are never suggested — nothing can be
    #: imported into them (plan section 3.5.8 / 13.1).
    suggested_mapping: dict[str, str | None]
    row_count: int
    #: First few rows, raw (pre-validation) cell text, for the Owner to
    #: sanity-check the mapping against real data before confirming it.
    sample_rows: list[dict[str, str]]


class ImportMappingRequest(BaseModel):
    """The shared body for `/validate` and `/commit` — the Owner's confirmed
    mapping plus how to handle duplicates and ambiguous dates."""

    #: CSV header -> column key. A header omitted here is simply not
    #: imported (its data is dropped, same as an unmapped column always
    #: would be).
    mapping: dict[str, str] = Field(default_factory=dict)
    #: Column keys whose combined value identifies "the same row" across an
    #: import and the page's existing data (e.g. Date + Cheque Number).
    #: Empty means every row is a new row (`duplicate_strategy` has nothing
    #: to act on).
    natural_key_columns: list[str] = Field(default_factory=list)
    duplicate_strategy: DuplicateStrategy = "create_anyway"
    date_formats: dict[str, DateFormatHint] = Field(default_factory=dict)


class ImportValidateResponse(BaseModel):
    total_rows: int
    valid_rows: int
    #: Rows that matched an existing natural key under `duplicate_strategy
    #: == "skip"` — counted separately from `error_rows` since these aren't
    #: wrong, just already present.
    duplicate_rows: int
    error_rows: int
    errors: list[ImportRowError]
    #: True when `errors` was capped (plan section 13.1's "per-row error
    #: list" — capped here at `MAX_IMPORT_ERRORS` rather than an unbounded
    #: response for a CSV with thousands of bad rows).
    errors_truncated: bool


class ImportCommitResponse(BaseModel):
    batch_id: uuid.UUID
    total_rows: int
    imported_rows: int
    skipped_rows: int
    error_rows: int


class ImportBatchOut(BaseModel):
    id: uuid.UUID
    page_id: uuid.UUID
    file_name: str
    total_rows: int
    imported_rows: int
    skipped_rows: int
    status: str
    created_at: datetime
    #: Computed server-side: `status == "COMMITTED"` and within 24 hours of
    #: `created_at` (plan section 13.1).
    can_rollback: bool


class ImportRollbackResponse(BaseModel):
    batch_id: uuid.UUID
    rolled_back_rows: int

File: apps/mobile/lib/features/pages/domain/csv_import.dart
/// CSV import (plan section 13.1, P3.5 Part 3) — mirrors
/// `app/schemas/csv_import.py`. Owner only, one generic pipeline for any
/// page: preview -> validate -> commit -> optional rollback within 24
/// hours.
library;

/// `POST /pages/{id}/import/preview`'s response — encoding/delimiter
/// detection, a fuzzy-suggested header -> column-key mapping, and a few
/// sample rows so the Owner can sanity-check the mapping before confirming
/// it.
class ImportPreview {
  const ImportPreview({
    required this.encoding,
    required this.delimiter,
    required this.headers,
    required this.suggestedMapping,
    required this.rowCount,
    required this.sampleRows,
  });

  factory ImportPreview.fromJson(Map<String, dynamic> json) {
    final rawMapping = (json['suggested_mapping'] as Map<String, dynamic>? ?? const {});
    return ImportPreview(
      encoding: json['encoding'] as String,
      delimiter: json['delimiter'] as String,
      headers: (json['headers'] as List<dynamic>).cast<String>(),
      suggestedMapping: rawMapping.map((k, v) => MapEntry(k, v as String?)),
      rowCount: json['row_count'] as int,
      sampleRows: (json['sample_rows'] as List<dynamic>)
          .cast<Map<String, dynamic>>()
          .map((row) => row.map((k, v) => MapEntry(k, v as String)))
          .toList(),
    );
  }

  final String encoding;
  final String delimiter;
  final List<String> headers;
  final Map<String, String?> suggestedMapping;
  final int rowCount;
  final List<Map<String, String>> sampleRows;
}

class ImportRowError {
  const ImportRowError({required this.row, this.column, required this.message});

  factory ImportRowError.fromJson(Map<String, dynamic> json) => ImportRowError(
    row: json['row'] as int,
    column: json['column'] as String?,
    message: json['message'] as String,
  );

  final int row;
  final String? column;
  final String message;
}

class ImportValidation {
  const ImportValidation({
    required this.totalRows,
    required this.validRows,
    required this.duplicateRows,
    required this.errorRows,
    required this.errors,
    required this.errorsTruncated,
  });

  factory ImportValidation.fromJson(Map<String, dynamic> json) => ImportValidation(
    totalRows: json['total_rows'] as int,
    validRows: json['valid_rows'] as int,
    duplicateRows: json['duplicate_rows'] as int,
    errorRows: json['error_rows'] as int,
    errors: (json['errors'] as List<dynamic>)
        .cast<Map<String, dynamic>>()
        .map(ImportRowError.fromJson)
        .toList(),
    errorsTruncated: json['errors_truncated'] as bool,
  );

  final int totalRows;
  final int validRows;
  final int duplicateRows;
  final int errorRows;
  final List<ImportRowError> errors;
  final bool errorsTruncated;
}

class ImportCommitResult {
  const ImportCommitResult({
    required this.batchId,
    required this.totalRows,
    required this.importedRows,
    required this.skippedRows,
    required this.errorRows,
  });

  factory ImportCommitResult.fromJson(Map<String, dynamic> json) => ImportCommitResult(
    batchId: json['batch_id'] as String,
    totalRows: json['total_rows'] as int,
    importedRows: json['imported_rows'] as int,
    skippedRows: json['skipped_rows'] as int,
    errorRows: json['error_rows'] as int,
  );

  final String batchId;
  final int totalRows;
  final int importedRows;
  final int skippedRows;
  final int errorRows;
}

/// How a row matching an existing natural key is handled (plan section
/// 13.1: "skip / update / create-anyway").
enum DuplicateStrategy {
  skip,
  update,
  createAnyway;

  String get wire => switch (this) {
    DuplicateStrategy.skip => 'skip',
    DuplicateStrategy.update => 'update',
    DuplicateStrategy.createAnyway => 'create_anyway',
  };

  String get label => switch (this) {
    DuplicateStrategy.skip => 'Skip duplicates',
    DuplicateStrategy.update => 'Update the existing record',
    DuplicateStrategy.createAnyway => 'Create a new record anyway',
  };
}

/// How to read an ambiguous (slash-separated) date column — set once per
/// column, never guessed (plan section 13.1).
enum DateFormatHint {
  ymd,
  dmy,
  mdy;

  String get wire => switch (this) {
    DateFormatHint.ymd => 'YMD',
    DateFormatHint.dmy => 'DMY',
    DateFormatHint.mdy => 'MDY',
  };

  String get label => switch (this) {
    DateFormatHint.ymd => 'Year / Month / Day',
    DateFormatHint.dmy => 'Day / Month / Year',
    DateFormatHint.mdy => 'Month / Day / Year',
  };
}

/// The Owner's confirmed choices for `/validate` and `/commit` — both take
/// exactly this shape (`ImportMappingRequest` server-side).
class ImportMapping {
  const ImportMapping({
    this.mapping = const {},
    this.naturalKeyColumns = const [],
    this.duplicateStrategy = DuplicateStrategy.createAnyway,
    this.dateFormats = const {},
  });

  final Map<String, String> mapping;
  final List<String> naturalKeyColumns;
  final DuplicateStrategy duplicateStrategy;
  final Map<String, DateFormatHint> dateFormats;

  ImportMapping copyWith({
    Map<String, String>? mapping,
    List<String>? naturalKeyColumns,
    DuplicateStrategy? duplicateStrategy,
    Map<String, DateFormatHint>? dateFormats,
  }) => ImportMapping(
    mapping: mapping ?? this.mapping,
    naturalKeyColumns: naturalKeyColumns ?? this.naturalKeyColumns,
    duplicateStrategy: duplicateStrategy ?? this.duplicateStrategy,
    dateFormats: dateFormats ?? this.dateFormats,
  );

  Map<String, dynamic> toJson() => {
    'mapping': mapping,
    'natural_key_columns': naturalKeyColumns,
    'duplicate_strategy': duplicateStrategy.wire,
    'date_formats': dateFormats.map((key, hint) => MapEntry(key, hint.wire)),
  };
}

class ImportBatch {
  const ImportBatch({
    required this.id,
    required this.pageId,
    required this.fileName,
    required this.totalRows,
    required this.importedRows,
    required this.skippedRows,
    required this.status,
    required this.createdAt,
    required this.canRollback,
  });

  factory ImportBatch.fromJson(Map<String, dynamic> json) => ImportBatch(
    id: json['id'] as String,
    pageId: json['page_id'] as String,
    fileName: json['file_name'] as String,
    totalRows: json['total_rows'] as int,
    importedRows: json['imported_rows'] as int,
    skippedRows: json['skipped_rows'] as int,
    status: json['status'] as String,
    createdAt: json['created_at'] as String,
    canRollback: json['can_rollback'] as bool,
  );

  final String id;
  final String pageId;
  final String fileName;
  final int totalRows;
  final int importedRows;
  final int skippedRows;
  final String status;
  final String createdAt;
  final bool canRollback;
}
